mlir/lib/Bindings/Python/IRCore.cpp

```diff
@@ -1249,6 +1249,19 @@ void PyOperationBase::writeBytecode(const py::object &fileObject,
                               .str());
 }
 
+void PyOperationBase::walk(py::object callback, bool usePreOrder) {
```

Comment:
I think it is possible to spell out the callback type as `std::function<void (MlirOperation)>` here, not sure if this will have any performance implications.

Consider exposing `MlirWalkResult` in C and Python API to allow functions to stop the walk early. 

I'm also not a fan of using a boolean kwarg for order, but can live with it.

Reply:
Thank you for suggestion!  I changed callback type to `std::function<MlirWalkResult (MlirOperation)>` as well as adding `MlirWalkResult` CAPI and python enums for `MlirWalkResult/MlirWalkOrder`.